docs/ADDING_MACHINES.md

```diff
@@ -78,14 +78,26 @@ MIMachineEvents.registerMachines(event => {
       items => items.addSlots(42, 27, 3, 3).addSlots(139, 27, 1, 3), fluids => {},
       /* MODEL CONFIGURATION */
       // front overlay?, top overlay?, side overlay?
-      true, true, false,
+      true, true, false
   );
 })
 ```
+If bronze or steel is in the list of tiers, the machine behavior defaults to allow gunpowder to double the speed of the machine. You can change this behavior by adding a config lambda to the end.
 
-## Adding an electric multiblock crafting machine
-These are the standard electric multiblock machines, such as the pressurizer, the electric quarry, etc...
+Example:
+```
+// steamCustomOverclock(Speed multiplier, Map of item or block to duration in ticks) 
+config => config.steamCustomOverclock(3, {"minecraft:redstone": 400, "minecraft:redstone_block": 3600 })
+```
```

Comment:
What if we want to allow different speed multipliers depending on the material? I feel like that would be a useful addition.

Reply:
That would be interesting but would change the default behavior a lot more. The default behavior of gunpowder is to allow stacking the tick duration as you add more gunpowder. If each catalyst were to affect the speed directly then it would have to overwrite the existing. Take for example the following hypothetical format
```
config.steamCustomOverclock({
	"minecraft:redstone": {
		"multiplier": 2,
		"ticks": 400
	},
	"minecraft:glowstone_dust": {
		"multiplier": 3,
		"ticks": 1000
	}
})
```
The intent is that glowstone is the more powerful catalyst and lasts longer than redstone.

If I right click with redstone in my hand twice, resulting in 800 ticks at 2x multiplier, what should happen if I right click with glowstone dust? Should it be at 3x multiplier with 1800 ticks allowing me to accumulate ticks from a different catalyst on a more powerful one? Or should it override and become 3x multiplier with 1000 ticks making me lose all my catalyst usages on the redstone?

What if I were to reverse the process and add a glowstone dust first and then redstone. Do I lose the 3x multiplier and reset back to 2x with 400 ticks? That would feel bad.

It's possible to implement either way but it would require retraining the player from the default behavior that throwing more catalysts in gets you more time.

Reply:
Perhaps another implementation could be to prevent the user from inserting a new catalyst until the existing one finishes? All of these could be correct depending on the design.

Reply:
Ah, I see what you mean, that's a very good point. I'll think more about it.

Reply:
I think we should keep track of all the remaining ticks for each possible multiplier, and we apply the fastest multiplier first.

For example if we have 10 ticks of x2 and we suddenly add 5 ticks of x3, it will be x3 for the first 5 ticks and then x2 for the remaining 10 ticks.

This can be done using a `SortedMap` (`TreeMap` implementation), as it allows getting the last entry of the map easily.

Reply:
It looks like it will have to be a `NavigableMap` since `SortedMap.lastKey()` throws if the map is empty (not something we should catch in tick) but `NavigableMap.lastEntry()` will return null if empty. It's assumed that multiplier entries with no ticks have to be removed from the map so getting the last entry updates to the next highest multiplier. Bear in mind lastKey or lastEntry with a TreeMap is still O(log N) called in tick.

Since a lookup for the current multiplier and a deduction in its tick count happens every tick and additions to ticks on a specific multiplier only happen when a catalyst is inserted (very infrequent) it would be more performant to use an `ArrayList`. Walk it on update or insert for O(N) to prevent duplicates and keep it sorted. Getting the highest multiplier in tick by just grabbing and updating or removing the last element is O(1).

Either implementation works but its your call. There's probably far worse things than O(log N) at 20hz going on in Minecraft to really matter. 🥲

Reply:
Ah yes array list works fine too of course. Whatever is fine by me :)

> Also with this addition it likely means the multiblock machine needs to display its current multiplier and remaining ticks to even know if its working correctly. Single block machines show the remaining time but not the multiplier.

We don't need to show too much info, this is an edge case after all. The current amount of info is probably fine.